rfcs/0024-nested-relationships.md

```diff
@@ -0,0 +1,577 @@
+# Nested Relationships
+
+## Problem
+
+Imagine the following data model, in which we have three collections `Invoice`, `Suburb` and `Campaign`. The `Invoice` type has a nested object type `Addresses` that has foreign keys to the `Address` collection. `Invoice` also has a nested array of `Discount` objects, of which each has a foreign key to the `Campaign` collection. So `Invoice` can be said to have "nested relationships" to both `Address` and `Campaign`.
```

Comment:
Can this capability enable self-joining within a document?

For example:


```json
{
   "dispatched_orders": [
        {"order_id": 1},
        {"order_id": 2}
   ],
   "orders": [
         {"order_id": 1, "product_name": "ABC", "ordered_at": "2024-10-15"},

         {"order_id": 2, "product_name": "DEF", "ordered_at": "2024-10-15"}
     ]
}
```

Considering the above data, I want to retrieve all the dispatched orders along with the order details. 

P.S: BTW, I realize this is a terrible example because the data model could have a key called `dispatched` within every order item. I will think of a better example (if I can think one of) and add it to this thread.

Reply:
Not _really_, no. You can join back to the same collection of documents, so if your `orders.order_id` property was unique across all documents in the collection (ie no two documents contained the same `orders.order_id`), then yes it would work. A relationship always targets a collection; that is not changing with this RFC.

Reply:
Actually, I take that back, it wouldn't work anyway, because `orders` is an array, and relationships can only target nested properties so long as they are nested objects, not nested arrays. We don't support implicit existential quantification over nested arrays at this time.

Reply:
Makes sense, thanks!